File: frontend/src/services/paymentService.ts
/**
 * Payment API service for payment logging and tracking
 */

import api from './api';

// Types
export interface PaymentMethod {
    type: 'bank_transfer' | 'check' | 'cash' | 'debit_card' | 'credit_card' | 'auto_pay' | 'other';
}

export interface PaymentCreate {
    debt_id: string;
    amount: number;
    payment_date?: string;
    payment_method?: string;
    notes?: string;
    new_balance?: number;
}

export interface PaymentUpdate {
    amount?: number;
    payment_date?: string;
    payment_method?: string;
    notes?: string;
    confirmed?: boolean;
}

export interface Payment {
    id: string;
    user_id: string;
    debt_id: string;
    plan_id?: string;
    amount: number;
    payment_date: string;
    payment_method?: string;
    confirmed: boolean;
    new_balance: number;
    interest_saved?: number;
    notes?: string;
    created_at: string;
    debt_name?: string;
}

export interface PaymentListResponse {
    payments: Payment[];
    total_count: number;
    total_amount: number;
}

export interface PaymentStats {
    total_payments: number;
    total_amount_paid: number;
    total_interest_saved: number;
    payments_this_month: number;
    amount_this_month: number;
    payments_last_30_days: number;
    amount_last_30_days: number;
    current_streak_days: number;
    longest_streak_days: number;
    on_track_percentage: number;
    average_payment_amount: number;
    payments_by_debt?: Record<string, number>;
}

export interface RecentPaymentSummary {
    last_payment_date?: string;
    last_payment_amount?: number;
    last_payment_debt_name?: string;
    payments_this_week: number;
    amount_this_week: number;
    total_principal_paid: number;
    total_interest_paid: number;
    current_streak: number;
    is_on_streak: boolean;
}

export interface MilestoneType {
    type: 'debt_paid_off' | 'percentage_milestone' | 'consistency_streak' | 'negotiation_success' | 'savings_milestone' | 'first_payment' | 'monthly_goal';
}

export interface Milestone {
    milestone_type: string;
    title: string;
    description: string;
    badge_name?: string;
    achievement_value?: number;
    related_debt_id?: string;
    related_debt_name?: string;
    celebration_message: string;
    interest_saved?: number;
    achieved_at: string;
}

export interface MilestoneCheckResult {
    milestones: Milestone[];
    has_new_milestones: boolean;
    celebration_priority: number;
}

export interface PaymentWithMilestones extends Payment {
    milestones?: MilestoneCheckResult;
}

// Daily Actions Types
export interface DailyAction {
    action_type: 'payment' | 'review' | 'rest' | 'milestone' | 'nudge';
    priority: 1 | 2 | 3 | 4 | 5;
    title: string;
    description: string;
    suggested_amount?: number;
    debt_id?: string;
    debt_name?: string;
    due_date?: string;
    is_overdue: boolean;
    motivational_message?: string;
}

export interface DailyActionsResponse {
    actions: DailyAction[];
    date: string;
    summary: string;
    progress_message?: string;
    streak_message?: string;
}

class PaymentService {
    /**
     * Log a new payment
     */
    async logPayment(data: PaymentCreate): Promise<PaymentWithMilestones> {
        const response = await api.post<PaymentWithMilestones>('/payments', data);
        return response.data;
    }

    /**
     * Get all payments with optional filters
     */
    async getPayments(options?: {
        debt_id?: string;
        limit?: number;
        start_date?: string;
        end_date?: string;
    }): Promise<PaymentListResponse> {
        const params = new URLSearchParams();
        if (options?.debt_id) params.append('debt_id', options.debt_id);
        if (options?.limit) params.append('limit', options.limit.toString());
        if (options?.start_date) params.append('start_date', options.start_date);
        if (options?.end_date) params.append('end_date', options.end_date);

        const response = await api.get<PaymentListResponse>(`/payments?${params.toString()}`);
        return response.data;
    }

    /**
     * Get recent payments
     */
    async getRecentPayments(days: number = 30, limit: number = 10): Promise<Payment[]> {
        const response = await api.get<Payment[]>(`/payments/recent?days=${days}&limit=${limit}`);
        return response.data;
    }

    /**
     * Get payment statistics
     */
    async getPaymentStats(): Promise<PaymentStats> {
        const response = await api.get<PaymentStats>('/payments/stats');
        return response.data;
    }

    /**
     * Get recent payment summary for dashboard
     */
    async getPaymentSummary(): Promise<RecentPaymentSummary> {
        const response = await api.get<RecentPaymentSummary>('/payments/summary');
        return response.data;
    }

    /**
     * Get a single payment by ID
     */
    async getPayment(paymentId: string): Promise<Payment> {
        const response = await api.get<Payment>(`/payments/${paymentId}`);
        return response.data;
    }

    /**
     * Update a payment
     */
    async updatePayment(paymentId: string, data: PaymentUpdate): Promise<Payment> {
        const response = await api.patch<Payment>(`/payments/${paymentId}`, data);
        return response.data;
    }

    /**
     * Delete a payment
     */
    async deletePayment(paymentId: string): Promise<void> {
        await api.delete(`/payments/${paymentId}`);
    }

    /**
     * Get daily actions
     */
    async getDailyActions(): Promise<DailyActionsResponse> {
        const response = await api.get<DailyActionsResponse>('/plans/actions/daily');
        return response.data;
    }

    /**
     * Check for new milestones
     */
    async checkMilestones(): Promise<MilestoneCheckResult> {
        const response = await api.get<MilestoneCheckResult>('/payments/milestones');
        return response.data;
    }
}

// Export PaymentCreate as LogPaymentRequest alias for compatibility
export type LogPaymentRequest = PaymentCreate;

export const paymentService = new PaymentService();
export default paymentService;

File: backend/app/services/payment_service.py
"""
Payment service for logging payments and managing payment-related operations.

Handles payment creation, debt balance updates, plan recalculation triggers,
and interest savings calculations.
"""

from datetime import date, datetime
from typing import Optional, List, Tuple

from app.models.payment import (
    PaymentCreate,
    PaymentUpdate,
    PaymentResponse,
    PaymentListResponse,
    PaymentStats,
    RecentPaymentSummary
)
from app.models.debt import DebtResponse
from app.db.repositories.payment_repo import PaymentRepository
from app.db.repositories.debt_repo import DebtRepository
from app.db.repositories.plan_repo import PlanRepository
from app.services.encryption_service import encryption_service
from app.agents.habit_agent import habit_agent, MilestoneCheckResult
from app.core.errors import ValidationError, NotFoundError


class PaymentService:
    """Service for payment operations."""
    
    @staticmethod
    async def log_payment(
        user_id: str,
        payment_data: PaymentCreate,
        token: Optional[str] = None
    ) -> Tuple[PaymentResponse, Optional[MilestoneCheckResult]]:
        """
        Log a new payment and update debt balance.
        
        Args:
            user_id: User UUID
            payment_data: Payment creation data
            token: Optional JWT token
        
        Returns:
            Tuple of (created payment, milestone check result if any)
        
        Raises:
            NotFoundError: If debt not found
            ValidationError: If payment amount exceeds balance
        """
        # 1. Get the debt to validate and calculate new balance
        debt = await DebtRepository.get_by_id(
            payment_data.debt_id,
            user_id,
            token=token
        )
        
        if not debt:
            raise NotFoundError(
                message="Debt not found",
                resource_type="debt",
                resource_id=payment_data.debt_id
            )
        
        if debt.is_paid_off:
            raise ValidationError(
                message="This debt is already paid off",
                field="debt_id"
            )
        
        # 2. Calculate new balance
        current_balance = debt.balance
        payment_amount = payment_data.amount
        
        if payment_amount > current_balance * 1.1:  # Allow 10% overpayment for rounding
            raise ValidationError(
                message=f"Payment amount (${payment_amount:.2f}) exceeds debt balance (${current_balance:.2f})",
                field="amount"
            )
        
        new_balance = max(0, current_balance - payment_amount)
        payment_data.new_balance = new_balance
        
        # 3. Calculate interest saved (simplified - based on APR and early payment)
        # More accurate calculation would need the full amortization schedule
        monthly_rate = (debt.apr / 100) / 12
        interest_saved = payment_amount * monthly_rate  # One month of interest on the principal paid
        
        # 4. Get active plan ID if exists
        active_plan = await PlanRepository.get_active_plan(user_id)
        plan_id = active_plan.id if active_plan else None
        
        # 5. Create the payment record
        payment = await PaymentRepository.create(
            user_id=user_id,
            payment_data=payment_data,
            plan_id=plan_id,
            interest_saved=interest_saved,
            token=token
        )
        
        # 6. Update debt balance
        debt_paid_off = new_balance == 0
        await DebtRepository.update_balance(
            debt_id=payment_data.debt_id,
            user_id=user_id,
            new_balance=new_balance,
            is_paid_off=debt_paid_off,
            token=token
        )
        
        # 7. Check for milestones
        milestone_result = None
        try:
            # Get payment stats for milestone checking
            payment_stats = await PaymentRepository.get_stats(user_id, token=token)
            
            # Get total debt info for percentage calculation
            debts_response = await DebtRepository.get_all_by_user(user_id, include_inactive=True)
            all_debts = debts_response.debts
            
            total_current = sum(d.balance for d in all_debts if d.is_active and not d.is_paid_off)
            total_original = sum(d.balance for d in all_debts)  # Simplified - would need original balances
            debts_paid_off_count = sum(1 for d in all_debts if d.is_paid_off)
            
            milestone_result = await habit_agent.check_milestones(
                user_id=user_id,
                total_debt_original=total_original + payment_stats.total_amount_paid,  # Approximate
                total_debt_current=total_current,
                total_paid=payment_stats.total_amount_paid,
                total_interest_saved=payment_stats.total_interest_saved,
                debts_paid_off=debts_paid_off_count,
                payment_stats=payment_stats,
                recently_paid_debt_id=payment_data.debt_id if debt_paid_off else None,
                recently_paid_debt_name=debt.creditor_name if debt_paid_off else None,
                existing_badges=[]  # Would need to track in DB
            )
        except Exception as e:
            # Don't fail the payment if milestone check fails
            print(f"Milestone check failed: {e}")
        
        # Add debt name to response
        payment.debt_name = debt.creditor_name
        
        return payment, milestone_result
    
    @staticmethod
    async def get_payment(
        payment_id: str,
        user_id: str,
        token: Optional[str] = None
    ) -> Optional[PaymentResponse]:
        """
        Get a single payment by ID.
        
        Args:
            payment_id: Payment UUID
            user_id: User UUID
            token: Optional JWT token
        
        Returns:
            Payment or None
        """
        return await PaymentRepository.get_by_id(payment_id, user_id, token=token)
    
    @staticmethod
    async def get_payments(
        user_id: str,
        debt_id: Optional[str] = None,
        limit: Optional[int] = None,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        token: Optional[str] = None
    ) -> PaymentListResponse:
        """
        Get payments for a user with optional filters.
        
        Args:
            user_id: User UUID
            debt_id: Optional filter by debt
            limit: Max payments to return
            start_date: Filter by payment_date >= start_date
            end_date: Filter by payment_date <= end_date
            token: Optional JWT token
        
        Returns:
            PaymentListResponse with payments and totals
        """
        if debt_id:
            payments = await PaymentRepository.get_by_debt(
                debt_id, user_id, limit=limit, token=token
            )
            return PaymentListResponse(
                payments=payments,
                total_count=len(payments),
                total_amount=sum(p.amount for p in payments)
            )
        
        return await PaymentRepository.get_by_user(
            user_id,
            limit=limit,
            start_date=start_date,
            end_date=end_date,
            token=token
        )
    
    @staticmethod
    async def get_recent_payments(
        user_id: str,
        days: int = 30,
        limit: int = 10,
        token: Optional[str] = None
    ) -> List[PaymentResponse]:
        """
        Get recent payments for a user.
        
        Args:
            user_id: User UUID
            days: Number of days to look back
            limit: Max payments to return
            token: Optional JWT token
        
        Returns:
            List of recent payments
        """
        return await PaymentRepository.get_recent(user_id, days=days, limit=limit, token=token)
    
    @staticmethod
    async def get_payment_stats(
        user_id: str,
        token: Optional[str] = None
    ) -> PaymentStats:
        """
        Get payment statistics for a user.
        
        Args:
            user_id: User UUID
            token: Optional JWT token
        
        Returns:
            PaymentStats with aggregated data
        """
        return await PaymentRepository.get_stats(user_id, token=token)
    
    @staticmethod
    async def get_recent_summary(
        user_id: str,
        token: Optional[str] = None
    ) -> RecentPaymentSummary:
        """
        Get recent payment summary for dashboard.
        
        Args:
            user_id: User UUID
            token: Optional JWT token
        
        Returns:
            RecentPaymentSummary
        """
        return await PaymentRepository.get_recent_summary(user_id, token=token)
    
    @staticmethod
    async def update_payment(
        payment_id: str,
        user_id: str,
        update_data: PaymentUpdate,
        token: Optional[str] = None
    ) -> Optional[PaymentResponse]:
        """
        Update a payment record.
        
        Args:
            payment_id: Payment UUID
            user_id: User UUID
            update_data: Fields to update
            token: Optional JWT token
        
        Returns:
            Updated payment or None
        """
        return await PaymentRepository.update(
            payment_id, user_id, update_data, token=token
        )
    
    @staticmethod
    async def delete_payment(
        payment_id: str,
        user_id: str,
        token: Optional[str] = None
    ) -> bool:
        """
        Delete a payment record.
        
        Note: This should also revert the debt balance change,
        but for now we'll just delete the record.
        
        Args:
            payment_id: Payment UUID
            user_id: User UUID
            token: Optional JWT token
        
        Returns:
            True if deleted
        """
        # Get the payment first to potentially revert balance
        payment = await PaymentRepository.get_by_id(payment_id, user_id, token=token)
        
        if not payment:
            raise NotFoundError(
                message="Payment not found",
                resource_type="payment",
                resource_id=payment_id
            )
        
        # Delete the payment
        deleted = await PaymentRepository.delete(payment_id, user_id, token=token)
        
        if deleted:
            # Revert the debt balance
            debt = await DebtRepository.get_by_id(payment.debt_id, user_id, token=token)
            if debt:
                new_balance = debt.balance + payment.amount
                await DebtRepository.update_balance(
                    debt_id=payment.debt_id,
                    user_id=user_id,
                    new_balance=new_balance,
                    is_paid_off=False,
                    token=token
                )
        
        return deleted

    @staticmethod
    async def check_milestones(
        user_id: str,
        token: Optional[str] = None
    ) -> MilestoneCheckResult:
        """
        Check for any achieved milestones for a user.
        
        Args:
            user_id: User UUID
            token: Optional JWT token
        
        Returns:
            MilestoneCheckResult with any achieved milestones
        """
        try:
            # Get payment stats
            payment_stats = await PaymentRepository.get_stats(user_id, token=token)
            
            # Get debt info
            debts_response = await DebtRepository.get_all_by_user(user_id, include_inactive=True)
            all_debts = debts_response.debts
            
            total_current = sum(d.balance for d in all_debts if d.is_active and not d.is_paid_off)
            total_original = sum(d.balance for d in all_debts)
            debts_paid_off_count = sum(1 for d in all_debts if d.is_paid_off)
            
            return await habit_agent.check_milestones(
                user_id=user_id,
                total_debt_original=total_original + payment_stats.total_amount_paid,
                total_debt_current=total_current,
                total_paid=payment_stats.total_amount_paid,
                total_interest_saved=payment_stats.total_interest_saved,
                debts_paid_off=debts_paid_off_count,
                payment_stats=payment_stats,
                recently_paid_debt_id=None,
                recently_paid_debt_name=None,
                existing_badges=[]
            )
        except Exception as e:
            print(f"Milestone check failed: {e}")
            return MilestoneCheckResult(
                milestones=[],
                has_new_milestones=False,
                celebration_priority=0
            )


# Convenience functions
async def log_payment(
    user_id: str,
    payment_data: PaymentCreate,
    token: Optional[str] = None
) -> Tuple[PaymentResponse, Optional[MilestoneCheckResult]]:
    """Log a new payment."""
    return await PaymentService.log_payment(user_id, payment_data, token=token)


async def get_payment_stats(user_id: str, token: Optional[str] = None) -> PaymentStats:
    """Get payment statistics."""
    return await PaymentService.get_payment_stats(user_id, token=token)
